File: backend-rust/src/model/search_filter.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchFilter {
    #[serde(default = "default_and")]
    pub operator: String,
    #[serde(default)]
    pub conditions: Vec<SearchFilter>,
    #[serde(default)]
    pub field: String,
    #[serde(default)]
    #[serde(rename = "type")]
    pub filter_type: String,
    #[serde(default)]
    pub value: Value,
}

fn default_and() -> String {
    "AND".to_string()
}

impl SearchFilter {
    pub fn eq(field: impl Into<String>, value: impl Into<Value>) -> Self {
        Self {
            operator: "AND".into(),
            field: field.into(),
            filter_type: "eq".into(),
            value: value.into(),
            ..Default::default()
        }
    }

    pub fn matches(record: &serde_json::Map<String, Value>, filter: &SearchFilter) -> bool {
        if !filter.conditions.is_empty() {
            if filter.operator.eq_ignore_ascii_case("OR") {
                return filter
                    .conditions
                    .iter()
                    .any(|c| Self::matches(record, c));
            }
            return filter
                .conditions
                .iter()
                .all(|c| Self::matches(record, c));
        }

        if filter.field.is_empty() || filter.filter_type.is_empty() {
            return true;
        }

        let actual = record.get(&filter.field);
        match actual {
            // field absent: treat as null for isnotnull/isnull, false for everything else
            None => match filter.filter_type.as_str() {
                "isnull" | "isNull" => true,
                "isnotnull" | "notNull" => false,
                _ => false,
            },
            Some(actual) => Self::evaluate_condition(actual, &filter.filter_type, &filter.value),
        }
    }

    fn evaluate_condition(actual: &Value, op: &str, expected: &Value) -> bool {
        match op {
            "eq" => {
                if actual == expected {
                    return true;
                }
                // Coerce number ↔ string (DB may store run=1 as integer or "1" as string)
                let as_f64 = |v: &Value| {
                    v.as_f64()
                        .or_else(|| v.as_str().and_then(|s| s.trim().parse::<f64>().ok()))
                };
                if let (Some(a), Some(e)) = (as_f64(actual), as_f64(expected)) {
                    return (a - e).abs() < f64::EPSILON;
                }
                false
            }
            "eqIgnoreCase" => {
                if let (Some(a), Some(e)) = (actual.as_str(), expected.as_str()) {
                    a.trim().eq_ignore_ascii_case(e.trim())
                } else {
                    false
                }
            }
            "ne" => actual != expected,
            "gt" => compare_values(actual, expected) > 0,
            "gte" => compare_values(actual, expected) >= 0,
            "lt" => compare_values(actual, expected) < 0,
            "lte" => compare_values(actual, expected) <= 0,
            "in" => expected
                .as_array()
                .map(|arr| arr.contains(actual))
                .unwrap_or(false),
            "notIn" => expected
                .as_array()
                .map(|arr| !arr.contains(actual))
                .unwrap_or(true),
            "like" => {
                if let Some(e) = expected.as_str() {
                    if e.is_empty() {
                        return true;
                    }
                    if let Some(a) = value_as_compare_str(actual) {
                        return a.to_lowercase().contains(&e.to_lowercase());
                    }
                }
                false
            }
            "prefix" => {
                if let (Some(a), Some(e)) = (actual.as_str(), expected.as_str()) {
                    a.to_lowercase().starts_with(&e.to_lowercase())
                } else {
                    false
                }
            }
            "regex" => {
                if let (Some(a), Some(e)) = (actual.as_str(), expected.as_str()) {
                    regex::Regex::new(e)
                        .map(|re| re.is_match(a))
                        .unwrap_or(false)
                } else {
                    false
                }
            }
            "range" => {
                if let Some(arr) = expected.as_array() {
                    if arr.len() == 2 {
                        return compare_values(actual, &arr[0]) >= 0
                            && compare_values(actual, &arr[1]) <= 0;
                    }
                }
                false
            }
            "isnotnull" | "notNull" => !matches!(actual, Value::Null),
            "isnull" | "isNull" => matches!(actual, Value::Null),
            // noteq: not equal, also handles string vs number like "1" vs 1
            "noteq" | "notEq" => {
                if actual == expected {
                    return false;
                }
                // also treat empty string as matching null for noteq
                if matches!(actual, Value::Null) {
                    return expected.as_str().map(|s| !s.is_empty()).unwrap_or(true);
                }
                true
            }
            _ => false,
        }
    }
}

fn compare_values(a: &Value, b: &Value) -> i32 {
    if let (Some(aa), Some(bb)) = (compare_as_number(a), compare_as_number(b)) {
        return aa
            .partial_cmp(&bb)
            .unwrap_or(std::cmp::Ordering::Equal) as i32;
    }
    if let (Some(aa), Some(bb)) = (a.as_str(), b.as_str()) {
        return aa.cmp(bb) as i32;
    }
    0
}

fn compare_as_number(v: &Value) -> Option<f64> {
    v.as_f64()
        .or_else(|| v.as_i64().map(|n| n as f64))
        .or_else(|| v.as_u64().map(|n| n as f64))
        .or_else(|| {
            v.as_str()
                .map(str::trim)
                .filter(|s| !s.is_empty())
                .and_then(|s| s.parse::<f64>().ok())
        })
}

fn value_as_compare_str(v: &Value) -> Option<String> {
    v.as_str().map(String::from).or_else(|| {
        v.as_i64()
            .map(|n| n.to_string())
            .or_else(|| v.as_u64().map(|n| n.to_string()))
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn compare_values_coerces_yyyymmdd_number_and_string() {
        assert!(compare_values(&json!("20250606"), &json!(20250606)) == 0);
        assert!(compare_values(&json!(20250607), &json!("20250606")) > 0);
        assert!(compare_values(&json!("20250605"), &json!(20250606)) < 0);
    }

    #[test]
    fn like_empty_matches_any_scalar_id() {
        let filter = SearchFilter {
            field: "id".into(),
            filter_type: "like".into(),
            value: json!(""),
            ..Default::default()
        };
        let mut row = serde_json::Map::new();
        row.insert("id".into(), json!(12345));
        assert!(SearchFilter::matches(&row, &filter));
    }

    #[test]
    fn field_ngay_range_matches_numeric_storage() {
        let filter = SearchFilter {
            operator: "AND".into(),
            conditions: vec![
                SearchFilter {
                    field: "field_ngay".into(),
                    filter_type: "gte".into(),
                    value: json!("20250601"),
                    ..Default::default()
                },
                SearchFilter {
                    field: "field_ngay".into(),
                    filter_type: "lte".into(),
                    value: json!("20250630"),
                    ..Default::default()
                },
            ],
            ..Default::default()
        };
        let mut row = serde_json::Map::new();
        row.insert("field_ngay".into(), json!(20250606));
        assert!(SearchFilter::matches(&row, &filter));
    }
}

File: backend-go/internal/model/search_filter.go
package model

import (
	"math"
	"strconv"
	"strings"
)

type SearchFilter struct {
	Operator   string         `json:"operator"`
	Conditions []SearchFilter `json:"conditions"`
	Field      string         `json:"field"`
	FilterType string         `json:"type"`
	Value      any            `json:"value"`
}

func EqFilter(field string, value any) SearchFilter {
	return SearchFilter{
		Operator:   "AND",
		Field:      field,
		FilterType: "eq",
		Value:      value,
	}
}

func (f SearchFilter) Type() string { return f.FilterType }

func (f SearchFilter) HasLike() bool {
	if strings.EqualFold(f.FilterType, "like") {
		if s, ok := asString(f.Value); ok && strings.TrimSpace(s) != "" {
			return true
		}
	}
	for _, c := range f.Conditions {
		if c.HasLike() {
			return true
		}
	}
	return false
}

func (f SearchFilter) CollectLikeTerms() []string {
	var out []string
	f.collectLikeTermsInto(&out)
	return out
}

func (f SearchFilter) collectLikeTermsInto(out *[]string) {
	if len(f.Conditions) > 0 {
		for _, c := range f.Conditions {
			c.collectLikeTermsInto(out)
		}
		return
	}
	if strings.EqualFold(f.FilterType, "like") {
		if s, ok := asString(f.Value); ok {
			s = strings.TrimSpace(s)
			if s != "" {
				*out = append(*out, s)
			}
		}
	}
}

func (f SearchFilter) ValueString() string {
	switch v := f.Value.(type) {
	case string:
		return strings.TrimSpace(v)
	case float64:
		return strconv.FormatInt(int64(v), 10)
	case int:
		return strconv.Itoa(v)
	case int64:
		return strconv.FormatInt(v, 10)
	default:
		if f.Value == nil {
			return ""
		}
		return strings.TrimSpace(stringify(f.Value))
	}
}

func (f SearchFilter) Matches(record map[string]any) bool {
	if len(f.Conditions) > 0 {
		if strings.EqualFold(f.Operator, "OR") {
			for _, c := range f.Conditions {
				if c.Matches(record) {
					return true
				}
			}
			return false
		}
		for _, c := range f.Conditions {
			if !c.Matches(record) {
				return false
			}
		}
		return true
	}
	if f.Field == "" || f.FilterType == "" {
		return true
	}
	actual, ok := record[f.Field]
	if !ok {
		switch f.FilterType {
		case "isnull", "isNull":
			return true
		case "isnotnull", "notNull":
			return false
		default:
			return false
		}
	}
	return evaluateCondition(actual, f.FilterType, f.Value)
}

func evaluateCondition(actual any, op string, expected any) bool {
	switch op {
	case "eq":
		if valuesEqual(actual, expected) {
			return true
		}
		a, okA := asFloat64(actual)
		e, okE := asFloat64(expected)
		if okA && okE {
			return math.Abs(a-e) < 1e-9
		}
		return false
	case "eqIgnoreCase":
		as, okA := asString(actual)
		es, okE := asString(expected)
		return okA && okE && strings.EqualFold(strings.TrimSpace(as), strings.TrimSpace(es))
	case "ne":
		return !valuesEqual(actual, expected)
	case "gt":
		return compareValues(actual, expected) > 0
	case "gte":
		return compareValues(actual, expected) >= 0
	case "lt":
		return compareValues(actual, expected) < 0
	case "lte":
		return compareValues(actual, expected) <= 0
	case "in":
		arr, ok := expected.([]any)
		if !ok {
			return false
		}
		for _, item := range arr {
			if valuesEqual(actual, item) {
				return true
			}
		}
		return false
	case "notIn":
		arr, ok := expected.([]any)
		if !ok {
			return true
		}
		for _, item := range arr {
			if valuesEqual(actual, item) {
				return false
			}
		}
		return true
	case "like":
		es, okE := asString(expected)
		if !okE || es == "" {
			return true
		}
		as, okA := compareString(actual)
		return okA && strings.Contains(strings.ToLower(as), strings.ToLower(es))
	case "prefix":
		as, okA := asString(actual)
		es, okE := asString(expected)
		return okA && okE && strings.HasPrefix(strings.ToLower(as), strings.ToLower(es))
	case "isnotnull", "notNull":
		return actual != nil
	case "isnull", "isNull":
		return actual == nil
	case "noteq", "notEq":
		if valuesEqual(actual, expected) {
			return false
		}
		if actual == nil {
			if es, ok := asString(expected); ok {
				return es != ""
			}
			return true
		}
		return true
	default:
		return false
	}
}

// ValuesEqual compares filter values with loose numeric/string equality (Java parity).
func ValuesEqual(a, b any) bool {
	return valuesEqual(a, b)
}

func valuesEqual(a, b any) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	switch av := a.(type) {
	case string:
		if bv, ok := b.(string); ok {
			return av == bv
		}
	case float64:
		if bv, ok := asFloat64(b); ok {
			return math.Abs(av-bv) < 1e-9
		}
	case bool:
		if bv, ok := b.(bool); ok {
			return av == bv
		}
	}
	return stringify(a) == stringify(b)
}

func compareValues(a, b any) int {
	af, okA := asFloat64(a)
	bf, okB := asFloat64(b)
	if okA && okB {
		switch {
		case af < bf:
			return -1
		case af > bf:
			return 1
		default:
			return 0
		}
	}
	as, okAS := compareString(a)
	bs, okBS := compareString(b)
	if okAS && okBS {
		return strings.Compare(as, bs)
	}
	return strings.Compare(stringify(a), stringify(b))
}

func asFloat64(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case string:
		f, err := strconv.ParseFloat(strings.TrimSpace(n), 64)
		return f, err == nil
	default:
		return 0, false
	}
}

func asString(v any) (string, bool) {
	if s, ok := v.(string); ok {
		return s, true
	}
	return "", false
}

func compareString(v any) (string, bool) {
	if s, ok := v.(string); ok {
		return s, true
	}
	if f, ok := asFloat64(v); ok {
		return strconv.FormatFloat(f, 'f', -1, 64), true
	}
	return "", false
}

func stringify(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	if f, ok := asFloat64(v); ok {
		return strconv.FormatFloat(f, 'f', -1, 64)
	}
	return ""
}
